### PR TITLE
refactor(ui): UI mechainsm rework

### PR DESCRIPTION
**Pull request by YvonTre**
# UI mechanism rework

## TL;DR

- Refactor UI logic: Move UI-related logic from `PlayingView` to `TheVideoController`.
  - Use UIManager: Cancel global event bus to pass ui-related variables.
  - Use TimerManager: Use custom `timerManager` to manage old seperate `setTimeout`s.
- Organize codes: Clean unnecessary codes for simplicity and rename some functions for better readability.
- Some UI tweaks: Optimize some UI logic, such as close popup when hitting another widgets, all widgets show/hide management and so-on.
- Unit tests: Rewrite TheVideoController's unit tests.

## Refactor UI logic

- Move all UI-related codes from `PlayingView` to `TheVideoController`.
- `TheVideoController` consists of three parts: playback/fullscreen control, UIManager and event handlers.
- Use custom `v-hidden` to update UI widget's final display state.

### Event handlers

- Event handlers mainly record necessary input information, like mousemove's position, keydown's keycode and other stuff.
- Current registered event handlers: `mousemove`, `mouseenter`/`mouseleave`, `mousedown`/`mouseup`, `dblclick`, `wheel` and `keydown`.
- Handler's name format: handle + eventType + eventAppendix, camelCase, like `handleMousemove`, `handleMousedownLeft`, `handleMouseleave`, `handleKeydown`.
- Most of events use Vue's `v-on`, but some global input like `keydown` and `wheel` use `document.addEventListener`.

### Playback/Fullscreen control

- Remove `isValidClick` (imported from #130 ) for performance issues.
- Use `mouseup.left` for playback control and `dblclick` for fullscreen control. 
- Use `preventSingleClick` and `clicksTimer` to avoid `dblclick` trigger `mouseup` handler.

### `UIManager`

- `UIManager` consists of four parts(period): `inputProcess`, `UITimerManager`, `UIDisplayManager`, `UIStateManager`.

#### `inputProcess`

- Convert recorded raw event information to different timer and state messages.
  - `mousemove` ---> `mouseStopMoving` timer
  - `mouseenter`/`mouseleave` ---> `mouseLeavingWindow` timer
  - `keydown`/`wheel` ---> `sleepingVolumeButton`/`sleepingProgressBar` timer
  - `mousedown` ---> `showAttached` state
- Also responsible for leaving lastEventInfo for next use.

#### `UITimerManager`

- Accept a `frameTime` indicates current frame's duration(in miliseconds).
- Convert real world's time passing to timerState.

#### `UIDisplayManager`

- Combine `showAllWidgets` and `timerState` altogether into UI widget's `displayState`.

#### `UIStateManager`

- Manage UI widgets state, currently there is only one state: `showAttached`.

### `v-hidden`

- `v-show` has a bug when it comes to `TimeProgressBar`.
- Want to manage UI widget's display/hide transition altogether.
- `v-hidden` is like `v-show`, but instead of setting `display` state, `v-hidden` set `visibility` to `hidden` or `visible`.
- `v-hidden` only apply style change when `binding.value` changed.

## Some UI tweaks

- When mouseover on any widget, all widgets will not disappear even if some keyboard events triggered.
- When all widgets is about to disappear and some keyboard/wheel events triggered, all widgets disappear timer will reset.
- When subtitle's menu is on and all widgets disappear, subtitle's menu is off when all widgets back to screen. The same applies to all widgets' attached part.

## To-do

- [ ] Finish `TheVideoController` unit test.
- [ ] Add comment for `TheVideoController`.
- [ ] Add benchmark test.

### CODE REVIEW

**Comment by huzhenghui37** [comment]
好长，我看过了，实际效果 OK

**Review by tomasen — APPROVED** [review]
